### PR TITLE
feat(changelog): add tag message to release context

### LINKED ISSUE #623: Allow including tag-message in changelog
### Is there an existing issue or pull request for this?

- [X] I have searched the existing issues and pull requests

### Feature description

To have a "headline" for a release, it would be great allow including tag-messages (of release tags) into the changelog. For example, a changelog including this headline could look like this:

```
# Changelog

All notable changes to this project will be documented in this file.

## [1.0.1] - 2021-07-18

This is the release-tag message

### Miscellaneous Tasks
- Add release script

### Refactor

- Expose string functions
```

### Desired solution

Provide variable (such as `tag_message`) to the context, so it can be used as a variable within the changelog template.

### Alternatives considered

N/A

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by MeitarR**
## Description

added to the release context the tag's message if it exists

## Motivation and Context

closes #623 

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (no code change)
- [X] Refactor (refactoring production code)
- [ ] Other 

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [x] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by orhun** [comment]
Looks good at first glance, I can add a proper review if you can fix the CI & polish the implementation 🐻

File: git-cliff/src/args.rs [hunk 1]
```diff
@@ -114,45 +114,54 @@ pub struct Opt {
...
 	#[arg(
 		long,
 		env = "GIT_CLIFF_WITH_COMMIT",
 		value_name = "MSG",
 		num_args(1..)
 	)]
-	pub with_commit:     Option<Vec<String>>,
+	pub with_commit:      Option<Vec<String>>,
+	/// Sets custom message to the latest release (will overwrite original tag's
+	/// message if exists).
```
Comment (orhun) [inline git-cliff/src/args.rs:146]: I think we should remove the message in parenthesis and mention this in the documentation instead. It makes the help output overly long.

File: git-cliff-core/src/tag.rs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+/// Common tag object that is parsed from a repository.
+/// lightweight tag will have None at message.
```
Comment (orhun) [inline git-cliff-core/src/tag.rs:2]:
```suggestion
/// Common tag object that is parsed from a repository.
///
/// Lightweight tags will have `None` as message.
```

File: git-cliff-core/src/repo.rs [hunk 3]
```diff
@@ -95,11 +96,47 @@ impl Repository {
...
+			.and_then(|describe| {
+				describe
+					.format(None)
+					.ok()
+					.map(|name| self.resolve_tag(&name))
+			})
+	}
+
+	/// Returns the tag object of the given name.
+	/// if given name don't exists, still returns Tag object with the given name
```
Comment (orhun) [inline git-cliff-core/src/repo.rs:112]:
```suggestion
	/// Returns the tag object of the given name.
	///
	/// If given name doesn't exist, it still returns `Tag` with the given name.
```

File: git-cliff-core/src/repo.rs [hunk 4]
```diff
@@ -95,11 +96,47 @@ impl Repository {
...
+			},
+		}
+	}
+
+	/// used to remove signature from signed tag message
+	fn cleanup_message(message: &str) -> String {
+		let re = Regex::new(
+			r"(?s)-----BEGIN PGP SIGNATURE-----(.*?)-----END PGP SIGNATURE-----",
+		)
+		.expect("invalid regex, wtf");
```
Comment (orhun) [inline git-cliff-core/src/repo.rs:137]: Let's use a lazy regex here.

e.g.

https://github.com/orhun/git-cliff/blob/7f594e04255a0ba697071e8e1b49e0f74d6ef4d9/git-cliff-core/src/commit.rs#L36

Then we can also remove `cleanup_message` function and simply use the regex in `resolve_tag`.

File: git-cliff-core/src/repo.rs [hunk 5]
```diff
@@ -95,11 +96,47 @@ impl Repository {
...
+					name:    name.to_owned(),
+					message: None,
+				},
+			},
+			_ => Tag {
+				name:    name.to_owned(),
+				message: None,
+			},
+		}
+	}
```
Comment (orhun) [inline git-cliff-core/src/repo.rs:130]:
```suggestion
	pub fn resolve_tag(&self, name: &str) -> Tag {
		match self
			.inner
			.resolve_reference_from_short_name(name)
			.and_then(|r| r.peel_to_tag())
		{
			Ok(tag) => Tag {
				name:    tag.name().unwrap_or_default().to_owned(),
				message: tag.message().map(Self::cleanup_message),
			},
			_ => Tag {
				name:    name.to_owned(),
				message: None,
			},
		}
	}
```

**Review by orhun — COMMENTED** [review]
Very nice! I had some comments.

Can you also add a little documentation under `website/docs/usage/adding-tag-messages.md`?

File: git-cliff-core/src/repo.rs (hunk 4 above)
Comment (MeitarR) [reply git-cliff-core/src/repo.rs]: cool! done

**Comment by MeitarR** [comment]
> Very nice! I had some comments.
> 
> Can you also add a little documentation under `website/docs/usage/adding-tag-messages.md`?

Thanks!

added the docs: eafe9b0

**Review by orhun — APPROVED** [review]
LGTM, TYSM!

**Comment by chrichrichri** [comment]
Thanks everybody for implementing this! @

**Comment by hedon954** [comment]
When will this feature be updated to git-cliff-action?

**Comment by MeitarR** [comment]
> When will this feature be updated to git-cliff-action?
> 

I don't use the github action (I use gitlab) so I'm not 100% sure. But, I guess it's already there and all you need is to use the `message` variable in the template. As described [in the docs](https://git-cliff.org/docs/usage/adding-tag-messages)

**Comment by orhun** [comment]
Just updated `git-cliff-action` to use `git-cliff` 2.4.0: https://github.com/orhun/git-cliff-action/releases/tag/v3.3.0

**Pull request merged**